pkg/azure/access/helpers/disk.go

```diff
@@ -32,7 +32,9 @@ const (
 // DeleteDisk deletes disk for passed in resourceGroup and diskName.
 // NOTE: All calls to this Azure API are instrumented as prometheus metric.
 func DeleteDisk(ctx context.Context, client *armcompute.DisksClient, resourceGroup, diskName string) (err error) {
-	defer instrument.RecordAzAPIMetric(err, diskDeleteServiceLabel, time.Now())
+	defer func(invocationTime time.Time) {
+		instrument.RecordAzAPIMetric(err, diskDeleteServiceLabel, invocationTime)
+	}(time.Now())
```

Comment:
Why now just create a utility function to assist here instead of repeating lambda dozens of times ?

```go
func APIMetricRecorder(err error, azServiceName string) func() {
 invocationTime := time.Now()
 return func() {
  instrument.RecordAzAPIMetric(err, diskDeleteServiceLabel, invocationTime)
 }
}
```

```go
defer MeasureTime(err, diskDeleteServiceLabel)
```

Reply:
ok thanks, will do that.

Reply:
PTAL. Changes are done for both API and Driver metrics

Reply:
This won't work. The error will still be nil because it is a parameter to the function and hence calculated before the execution and not at the instance the defer is executed.

Reply:
Yikes, this is right.🙈 `err` is a interface pointer in Golang which is treated as fat pointer copied by value, not reference, so the value will be `nil` at the end of the function even if was set midway. 
 
One needs to use a pointer to pointer approach here, sorry! (hitting my head). (Also, quickly tested this with/without error)

```go
func APIMetricRecorderFn(azServiceName string, errPtr *error) func() {
	invocationTime := time.Now()
	return func() {
		RecordAzAPIMetric(*errPtr, azServiceName, invocationTime)
	}
}
```

```go

func hailGod() error {
	var err error
	defer APIMetricRecorderFn("hailgod", &err)()
	fmt.Println("God Service Called")
	err = errors.New("noreply")
	return err
}

```
God Service Called
RecordAzAPIMetric: err: noreply, operation:hailgod, invocationTime: 2024-02-14 22:32:39.303707 +0530 IST m=+0.000132209

```

Reply:
good catch. Thanks. Using pointer to a pointer is not used very often in golang code. There are reasons for that but in this there does not seem to be side effect therefore i have made the proposed change.